### PR TITLE
Added change proposal support

### LINKED ISSUE #21: Change proposal support

### PR DESCRIPTION
**Pull request by Daniilmass**
Resolves #21

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -736,7 +736,236 @@ paths:
...
               schema:
                 $ref: "#/components/schemas/Error"
-
+  /v1/projects/{projectId}/milestones/{milestoneId}/tasks/{taskId}/changeProposals:
+    get:
+      operationId: listChangeProposals
+      summary: Get all change proposals for a milestone task
+      description: Get all change proposals for a milestone task
+      tags:
+        - Change Proposals
```
Comment (kaeline) [inline swagger.yaml]: Tags should be one word i think for the code generation

File: swagger.yaml [hunk 2]
```diff
@@ -1343,6 +1572,52 @@ components:
         - IN_PROGRESS
         - DONE
 
+    ChangeProposal:
```
Comment (kaeline) [inline swagger.yaml:1573]: Where does this object keep the actual proposed change? E.g new task object with changed fields?

File: swagger.yaml [hunk 3]
```diff
@@ -1343,6 +1572,52 @@ components:
...
+          format: uuid
+          description: The ID of the task associated with this change proposal
+        date:
+          type: string
+          format: date-time
+          description: The date of the change proposal
+        userId:
+          type: string
+          format: uuid
+          description: The ID of the user who created the change proposal
```
Comment (kaeline) [inline swagger.yaml]: We could just use metadata for this purposes?

**Review by kaeline — APPROVED** [review]

File: swagger.yaml [hunk 4]
```diff
@@ -736,7 +736,211 @@ paths:
...
+          schema:
+            type: integer
+        - name: max
+          in: query
+          description: Max results.
+          schema:
+            type: integer
+      responses:
+        "200":
+          description: List of change proposals for the milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
          description: List of change proposals
```

File: swagger.yaml [hunk 5]
```diff
@@ -736,7 +736,211 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Error"
-
+  /v1/projects/{projectId}/milestones/{milestoneId}/changeProposals:
+    get:
+      operationId: listChangeProposals
+      summary: Get all change proposals for a milestone task sort from old to new
+      description: Get all change proposals for a milestone task sort from old to new
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
      summary: List change proposals within a milestone
      description: Lists change proposals within a milestone. Proposals are sorted from earliest to latest created.
```

File: swagger.yaml [hunk 6]
```diff
@@ -736,7 +736,211 @@ paths:
...
+          description: Invalid request was sent to the server
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+  /v1/projects/{projectId}/milestones/{milestoneId}/changeProposals/{changeProposalId}:
+    get:
+      operationId: findChangeProposal
+      summary: Get a specific change proposal for a milestone task
+      description: Get a specific change proposal for a milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
      summary: Find a specific change proposal within a milestone
      description: Find a specific change proposal within a milestone
```

File: swagger.yaml [hunk 7]
```diff
@@ -736,7 +736,211 @@ paths:
...
+        "default":
+          description: Invalid request was sent to the server
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+    put:
+      operationId: updateChangeProposal
+      summary: Update a specific change proposal for a milestone task
+      description: Update a specific change proposal for a milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
      summary: Update a specific change proposal within a milestone
      description: Update a specific change proposal within a milestone
```

File: swagger.yaml [hunk 8]
```diff
@@ -736,7 +736,211 @@ paths:
...
+        "default":
+          description: Invalid request was sent to the server
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+    post:
+      operationId: createChangeProposal
+      summary: Create a new change proposal for a milestone task
+      description: Create a new change proposal for a milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
      summary: Create a new change proposal within a milestone
      description: Create a new change proposal within a milestone
```

File: swagger.yaml [hunk 9]
```diff
@@ -736,7 +736,211 @@ paths:
...
+            format: uuid
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ChangeProposal'
+      responses:
+        "200":
+          description: Updated change proposal for the milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
          description: Updated change proposal
```

File: swagger.yaml [hunk 10]
```diff
@@ -736,7 +736,211 @@ paths:
...
+            format: uuid
+        - name: changeProposalId
+          in: path
+          required: true
+          schema:
+            type: string
+            format: uuid
+      responses:
+        "200":
+          description: Specific change proposal for the milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
          description: Found change proposal
```

File: swagger.yaml [hunk 11]
```diff
@@ -736,7 +736,211 @@ paths:
...
+            format: uuid
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ChangeProposal'
+      responses:
+        "200":
+          description: Created change proposal for the milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
          description: Created change proposal
```

File: swagger.yaml [hunk 12]
```diff
@@ -736,7 +736,211 @@ paths:
...
+        "default":
+          description: Invalid request was sent to the server
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+    delete:
+      operationId: deleteChangeProposal
+      summary: Delete a specific change proposal for a milestone task
+      description: Delete a specific change proposal for a milestone task
```
Comment (lauriLukkarinen) [inline swagger.yaml]:
```suggestion
      summary: Delete a specific change proposal within a milestone
      description: Deletes a specific change proposal within a milestone
```

File: swagger.yaml [hunk 13]
```diff
@@ -1366,6 +1570,50 @@ components:
         - IN_PROGRESS
         - DONE
 
+    ChangeProposal:
+      type: object
+      description: Change proposal
+      required:
+        - status
+        - taskId
```
Comment (lauriLukkarinen) [inline swagger.yaml:1578]: Should reason also be required?
```suggestion
      required:
        - status
        - taskId
        - reason
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**